include/sys/onoff.h

```diff
@@ -0,0 +1,611 @@
+ * operation notification.
+ *
+ * Clients that use this initialization will by notified of the
+ * completion of operations submitted through on-off service API
+ * through the provided callback.  Note that callbacks may be invoked
+ * from various contexts depending on the specific service; see
+ * `onoff_client_callback`.
+ *
+ * @param cli pointer to the client state object.
+ *
+ * @param handler a function pointer to use for notification.
+ *
+ * @param user_data an opaque pointer passed to the handler to provide
+ * additional context.
+ */
+static inline void onoff_client_init_callback(struct onoff_client *cli,
+					      onoff_client_callback handler,
+					      void *user_data)
+{
+	__ASSERT_NO_MSG(cli != NULL);
+	__ASSERT_NO_MSG(handler != NULL);
+
+	*cli = (struct onoff_client){
+		.async = {
+			.callback = {
+				.handler = handler,
+				.user_data = user_data,
+			},
+		},
+		.flags = ONOFF_CLIENT_NOTIFY_CALLBACK,
+	};
+}
+
+/**
+ * @brief Request a reservation to use an on-off service.
+ *
+ * The return value indicates the success or failure of an attempt to
+ * initiate an operation to request the resource be made available.
+ * If initiation of the operation succeeds the result of the request
+ * operation is provided through the configured client notification
+ * method, possibly before this call returns.
+ *
+ * Note that the call to this function may succeed in a case where the
+ * actual request fails.  Always check the operation completion
+ * result.
+ *
+ * @param srv the service that will be used.
+ *
+ * @param cli a non-null pointer to client state providing
+ * instructions on synchronous expectations and how to notify the
+ * client when the request completes.  Behavior is undefined if client
+ * passes a pointer object associated with an incomplete service
+ * operation.
+ *
+ * @retval Non-negative on successful (initiation of) request
+ * @retval -EIO if service has recorded an an error
+ * @retval -EINVAL if the parameters are invalid
+ * @retval -EAGAIN if the reference count would overflow
+ * @retval -EWOULDBLOCK if the function was invoked from non-thread
+ * context and successful initiation could result in an attempt to
+ * make the calling thread sleep.
+ */
+__syscall int onoff_request(struct onoff_service *srv,
+			    struct onoff_client *cli);
+
+/**
+ * @brief Release a reserved use of an on-off service.
+ *
+ * The return value indicates the success or failure of an attempt to
+ * initiate an operation to release of the resource.  If initiation of
+ * the operation succeeds the result of the release operation itself
+ * is provided through the configured client notification method,
+ * possibly before this call returns.
+ *
+ * Note that the call to this function may succeed in a case where the
+ * actual release fails.  Always check the operation completion
+ * result.
+ *
+ * Clients will be notified with `-EWOULDBLOCK` without causing a
```

Comment:
I find this hard to read.

Would this work?

```
 * Clients will be notified with `-EWOULDBLOCK` without causing a
 * service error if the attempt to release the resource would result in
 * a call to a stop transition that would sleep if the release call is
 * made from ISR or pre-kernel context.
```

Reply:
> I find this hard to read.

I still do

Reply:
This one went away because it can't happen.  I replaced the one in `onoff_request()`; please follow-up there.